### PR TITLE
Make `asarray` behavior consistent with Python Array API.

### LINKED ISSUE #70591: `torch.asarray` does not detect dtype of Python scalars
```python
import torch

for obj in (True, 1, 1.0):
    print(f"{obj} -> {torch.asarray(obj).dtype}")
```

```
True -> torch.float32
1 -> torch.float32
1.0 -> torch.float32
```

The [documentation](https://pytorch.org/docs/master/generated/torch.asarray.html) does not mention support for Python scalars, but since the function doesn't bail out, the behavior shown above is at least unexpected. 

`torch.asarray` was added in #60627 and specifically mentions the array API standard for compatibility, but this [defines behavior for scalars](https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray#asarray-obj-dtype-none-device-none-copy-none).

cc @gchanan @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi

### PR DESCRIPTION
**Pull request by ysiraichi**
Fixes #70591

This PR makes `torch.asarray` consistent with [the Python Array API](https://data-apis.org/array-api/latest/API_specification/generated/signatures.creation_functions.asarray.html#signatures.creation_functions.asarray) (which also happens to be the same as `torch.as_tensor` behavior). Specifically, it makes `asarray` casting conditional to the presence of the `dtype` argument. This solves the issue when Python scalars (and lists) were passed as input without specifying the `dtype`.

Before:
```python
>>> torch.asarray([True, False])
tensor([1., 0.])
```

After:
```python
>>> torch.asarray([True, False])
tensor([True, False])
```

### CODE REVIEW

File: torch/csrc/utils/tensor_new.cpp [hunk 1]
```diff
@@ -1163,8 +1163,11 @@ Tensor asarray(
     // Make tensor from sequence, inferring its type, and then convert
     // it to the desired type.
     tensor = internal_new_from_data(
-        TensorOptions(), dtype_unwrapped, device, obj, false, false, true);
-    tensor = tensor.to(dtype_unwrapped);
+        TensorOptions(), dtype_unwrapped, device, obj,
+        /* copy_variables = */ false, /* copy_numpy = */ false, /* type_inference = */ true);
+    if (dtype.has_value()) {
+      tensor = tensor.to(dtype_unwrapped);
+    }
```
Comment (Lezcano) [inline torch/csrc/utils/tensor_new.cpp:1170]: Could this be done by passing hte `dtype` within `TensorOptions()` to avoid the extra copy?

**Review by lezcano — APPROVED** [review]
Small comment, otherwise LGTM

File: test/test_tensor_creation_ops.py [hunk 2]
```diff
@@ -4033,6 +4033,27 @@ def check(**kwargs):
         check(requires_grad=False)
         check(requires_grad=False, copy=True)
 
+    @onlyCPU
+    def test_astensor_consistency(self, device):
```
Comment (mruberry) [inline test/test_tensor_creation_ops.py:4037]: Cool test

File: torch/csrc/utils/tensor_new.cpp [hunk 3]
```diff
@@ -1163,8 +1163,11 @@ Tensor asarray(
     // Make tensor from sequence, inferring its type, and then convert
     // it to the desired type.
     tensor = internal_new_from_data(
-        TensorOptions(), dtype_unwrapped, device, obj, false, false, true);
-    tensor = tensor.to(dtype_unwrapped);
+        TensorOptions(), dtype_unwrapped, device, obj,
+        /* copy_variables = */ false, /* copy_numpy = */ false, /* type_inference = */ true);
```
Comment (mruberry) [inline torch/csrc/utils/tensor_new.cpp:1167]: Thank you for clarifying this call with the updated comments

File: test/test_tensor_creation_ops.py (hunk 2 above)
Comment (mruberry) [reply test/test_tensor_creation_ops.py:4037]: Let's add a comment referencing the issue this PR fixes so people know what this test is doing and can see the history of why it was added

**Review by mruberry — COMMENTED** [review]
This is looking good, @ysiraichi! Should we also update the documentation to explain how asarray works when given a scalar? See here:

https://github.com/pytorch/pytorch/blob/dea61e7e6cad2022f7a1030a2b4bc084ce9afbc7/torch/_torch_docs.py#L992

(rendered page here: https://pytorch.org/docs/master/generated/torch.asarray.html?highlight=asarray#torch.asarray)

We should fix the `asarray` documentation while we're here, too:

"
Converts `obj` in a tensor.

`obj` can be one of:

1. a tensor
2. a NumPy array
3. a DLPack capsule
4. an object that implements Python's buffer protocol
5. a scalar
6. a sequence of scalars

When `obj` is a tensor, NumPy array, or DLPack capsule the returned tensor will, by default,  not require a gradient, have the same datatype as `obj`, be on the same device, and share memory with it. These properties can be controlled with the dtype, device, copy, and requires_grad keyword arguments. If the returned tensor is of a different datatype, on a different device, or a copy is requested then it will not share its memory with `obj`. If requires_grad is True then the returned tensor will require a gradient, and if `obj` is also a tensor with an autograd history then the returned tensor will have the same history.

When `obj` is not a tensor, NumPy Array, or DLPack capsule but implements Python's buffer protocol then the buffer is interpreted as an array of bytes grouped according to the size of the datatype passed to the dtype kwarg. (If no datatype is passed then the default floating point datatype is used, instead.) The returned tensor will have the specified datatype (or default floating point datatype if none is specified) and, by default, be on the CPU device and share memory with the buffer.

When `obj` is none of the above but a scalar or sequence of scalars then the returned tensor will, by default, infer its datatype from the scalar values, be on the CPU device, and not share its memory."

How does that sound? Does that look correct?

One thing that's interesting about this is that `torch.asarray([1, 2, 3])` currently returns a float32 tensor by default. Is that intended with this change, or should that return a long tensor? `torch.tensor((1, 2, 3))` returns a long tensor, by comparison.

The above documentation update should let us delete the paragraph starting with "For example: ..." The "See Also" note also needs it's formatting improved: each other function should appear on its own line, like the documentation for linalg.svd (https://pytorch.org/docs/master/generated/torch.linalg.svd.html?highlight=linalg%20svd#torch.linalg.svd). 

The argument and keyword argument documentation also needs an update:

"obj (object) -- a tensor, NumPy array, DLPack Capsule, object that implements Python's buffer protocol, scalar, or sequence of scalars."

"dtype (torch.dtype, optional) - the datatype of the returned tensor. Default: None, which causes the datatype of the returned tensor to be inferred from `obj`."
"copy (bool, optional) - controls whether the returned tensor shares memory with `obj`. If None, the default, then the returned tensor shares memory with `obj` whenever possible. If `True` then the returned tensor does not share its memory. If `False` then the returned tensor shares its memory with `obj` and an error is thrown if it cannot."
"device (torch.device, optional) - the device of the returned tensor. If None, the default, then the device of `obj` is used."
"requires_grad (bool, optional) - whether the returned tensor requires grad. If False, the default, then the returned tensor does not require a gradient. If requires_grad is True then the returned tensor will require a gradient, and if `obj` is also a tensor with an autograd history then the returned tensor will have the same history."

I'm curious to hear your thoughts, @ysiraichi, especially about how we should handle sequences of scalars.

File: torch/csrc/utils/tensor_new.cpp (hunk 1 above)
Comment (ysiraichi) [reply torch/csrc/utils/tensor_new.cpp:1170]: I don't think so. `internal_new_from_data` does not use its dtype (maybe it could swallow the dtype parameter in another PR, though). That said, we can make that extra copy go away by `type_inference = !dvalue.has_value()`.

Comment (Lezcano) [reply torch/csrc/utils/tensor_new.cpp:1170]: I was talking about the extra copy when calling `to`, but I guess you have a point as well with that.

So, doing `type_inference = !dvalue.has_value()` you would not need the `if`  and the `to`, right?
That way we would go from (potentially) 2 copies to just one, which is neat.

Comment (ysiraichi) [reply torch/csrc/utils/tensor_new.cpp:1170]: Yes, exactly!

**Comment by ysiraichi** [comment]
> How does that sound? Does that look correct?

I think that that looks correct.
It also sounds much better than what we have now. In summary, I think it's more clear and descriptive. Let's go with that!

> One thing that's interesting about this is that `torch.asarray([1, 2, 3])` currently returns a float32 tensor by default. Is that intended with this change, or should that return a long tensor? `torch.tensor((1, 2, 3))` returns a long tensor, by comparison.

That was indeed true. Not anymore, though. This PR also fixes that. It should be consistent with not only `as_tensor`, but whoever also uses `internal_new_from_data`.

I will update the documentation!

**Comment by ysiraichi** [comment]
@mruberry Do you have some time to take a look at this PR?

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -993,60 +993,71 @@ def merge_dicts(*dicts):
            r"""
 asarray(obj, *, dtype=None, device=None, copy=None, requires_grad=False) -> Tensor
 
-Converts :attr:`obj` into a tensor, sharing data and preserving autograd history
-if possible.
+Converts :attr:`obj` in a tensor.
```
Comment (anjali411) [inline torch/_torch_docs.py:996]: nit - Converts :attr:`obj` to a tensor.

File: torch/_torch_docs.py [hunk 5]
```diff
@@ -993,60 +993,71 @@ def merge_dicts(*dicts):
...
 
 :attr:`obj` can be one of:
 
 1. a tensor
 2. a NumPy array
 3. a DLPack capsule
-4. a Python object that implements the buffer protocol
-5. a Python sequence
+4. an object that implements Python's buffer protocol
+5. a sequence
```
Comment (anjali411) [inline torch/_torch_docs.py:1004]: does this just mean an ordered set?

File: torch/_torch_docs.py (hunk 4 above)
Comment (mruberry) [reply torch/_torch_docs.py:996]: My mistake! Sorry about that, @ysiraichi

File: torch/_torch_docs.py (hunk 5 above)
Comment (mruberry) [reply torch/_torch_docs.py:1004]: https://docs.python.org/3/library/stdtypes.html#sequence-types-list-tuple-range

This may be incorrect if we don't handle ranges. If we don't, however, we should update sequence to be "list" or "tuple" explicitly.

File: torch/_torch_docs.py [hunk 6]
```diff
@@ -993,60 +993,71 @@ def merge_dicts(*dicts):
...
-    device (:class:`torch.device`, optional): the device of the constructed tensor.
-           If `None`, then the device of :attr:`obj` is used. Else, it either copies
-           the data, if :attr:`obj` lives in a different device, or it shares the
-           memory, if :attr:`obj` lives in the same device.
-    requires_grad (bool, optional): If autograd should record operations on the
-           returned tensor. However, if this flag is ``False`` and the input object
-           is a non-leaf :class:`Tensor`, this function will call :func:`torch.Tensor.detach`.
+    dtype (:class:`torch.dtype`, optional): the datatype of the returned tensor.
+           Default: ``None``, which causes the datatype of the returned tensor to be
+           inferred from :attr:`obj`.
```
Comment (anjali411) [inline torch/_torch_docs.py:1048]: technically this affects the copy aspect of the behavior as mentioned above. might be worth re-iterating it here

**Review by anjali411 — APPROVED** [review]
LGTM overall! will let @mruberry take another look

File: torch/_torch_docs.py (hunk 5 above)
Comment (ysiraichi) [reply torch/_torch_docs.py:1004]: Oops. This looks like my fault, here. It should read "a scalar".
But, regarding ranges, I think we do support them.
```python
>>> torch.tensor(range(5))
tensor([0, 1, 2, 3, 4])
>>> torch.asarray(range(5))
tensor([0, 1, 2, 3, 4])
```

File: torch/_torch_docs.py (hunk 6 above)
Comment (ysiraichi) [reply torch/_torch_docs.py:1048]: Hmm. Not sure how I can do that in few words, since it affects the result depending on the input (already explained in the function summary). For example, for objects that implement Python's buffer protocol, `dtype` indicates how the sequence of bytes are going to be interpreted (no copy will take place). What do you think?

Comment (anjali411) [reply torch/_torch_docs.py:1048]: you can optionally add: 

As mentioned above, if :attr:`dtype` is not the same as :attr:`obj.dtype`, the returned tensor will not share its memory with :attr:`obj`

Comment (ysiraichi) [reply torch/_torch_docs.py:1048]: That sounds good. But, do you think it is okay adding it, even though it's not necessarily true for all cases?
I mean, it's not necessarily true if:
1. `obj` is something that implements Python's Buffer Protocol (memory will be shared no matter the value of `dtype`)
2. `obj` is a Python scalar or list of scalars (memory won't be shared no matter the value of `dtype`)
Let me know what you think!

Comment (anjali411) [reply torch/_torch_docs.py:1048]: oh I totally missed that! no let's skip then

**Comment by anjali411** [comment]
@ysiraichi this PR looks great! just needs a rebase before I can merge it

**Comment by anjali411** [comment]
@mruberry should we cherry-pick this for 1.11?

File: torch/_torch_docs.py [hunk 7]
```diff
@@ -993,60 +993,71 @@ def merge_dicts(*dicts):
...
 :attr:`obj` can be one of:
 
 1. a tensor
 2. a NumPy array
 3. a DLPack capsule
-4. a Python object that implements the buffer protocol
-5. a Python sequence
+4. an object that implements Python's buffer protocol
+5. a scalar
+6. a sequence of scalars
```
Comment (mruberry) [inline torch/_torch_docs.py:1005]: I think we still have some work to do here, @ysiraichi. Would you do me a favor and look at the following and we can file a follow-up PR if we need to make some changes?

For point 6 here, as mentioned elsewhere, does passing a range actually work?

File: torch/_torch_docs.py [hunk 8]
```diff
@@ -993,60 +993,71 @@ def merge_dicts(*dicts):
...
+
+When :attr:`obj` is none of the above but a scalar or sequence of scalars then the
+returned tensor will, by default, infer its datatype from the scalar values, be on the
+CPU device, and not share its memory.
 
-For each of the mentioned options, in order, this functions will assume :attr:`obj`
-is of that type and try, first, sharing memory. Only then, it will make a copy (if
-necessary).
+.. seealso::
+    :func:`torch.as_tensor` creates a tensor that always shares memory if the input is a
```
Comment (mruberry) [inline torch/_torch_docs.py:1028]: I think this needs a slight language tweak because the docs for `as_tensor` state:

```
Converts data into a tensor, sharing data and preserving autograd history if possible.
```

So I don't think as_tensor always creates a tensor that tensor that shares its memory. I would just delete this see also for simplicity because the difference between asarray and as_tensor is pretty subtle.

File: test/test_tensor_creation_ops.py [hunk 9]
```diff
@@ -4026,6 +4026,33 @@ def check(**kwargs):
...
+            [True, True, False],
+            [1, 2, 3, 42],
+            [0.0, 1.0, 2.0, 3.0],
+            # Mixed Lists
+            [True, False, 0],
+            [0.0, True, False],
+            [0, 1.0, 42],
+            [0.0, True, False, 42],
+            # With Complex
+            [0.0, True, False, 42, 5j],
```
Comment (mruberry) [inline test/test_tensor_creation_ops.py:4048]: Would you add a range object to this list of examples, too? I think ranges will work (but see comment below).

**Comment by mruberry** [comment]
Hey @ysiraichi, sorry for the delay getting back to you. Just have a few inline comments for your review. Since this PR is merged would you take a look and, if you agree with the comments, just open a tiny new PR addressing them? Then I can stamp it in for you.

File: torch/_torch_docs.py (hunk 7 above)
Comment (ysiraichi) [reply torch/_torch_docs.py:1005]: Yes, it does work with ranges also.

File: torch/_torch_docs.py (hunk 8 above)
Comment (ysiraichi) [reply torch/_torch_docs.py:1028]: Right. I guess the thing I wanted to convey here is that: the user has no control over whether to copy or not.

Comment (ysiraichi) [reply torch/_torch_docs.py:1028]: So, do you mean to remove the whole `see also` section, or just the entry for `as_tensor`?
